lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,472 @@
+  Sample data(values.getDimension(), 2);
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = k + 1.0;
+    data(k, 1) = values[k];
+  }
+
+  Cloud cloud(data, "red", "circle", "");
+  graph.add(cloud);
+
+  /* Min & max indices values */
+  const Scalar minInd = data.getMin()[1];
+  const Scalar maxInd = data.getMax()[1];
+
+
+  /* Add text description */
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = (k + 1.0) + 0.08;
+    data(k, 1) = values[k];
+  }
+
+  Description names(values.getDimension());
+  for(UnsignedInteger i = 0; i < values.getDimension(); ++i)
+  {
+    OSS oss;
+    oss << "X" << i + 1;
+    names[i] = String(oss);
+  }
+
+  Text text(data, names, "right");
+  text.setColor("black");
+  graph.add(text);
+
+  // Set bounding box
+  const Scalar step = maxInd - minInd;
+
+  Point lowerBound(2);
+  lowerBound[0] = 0.9;
+  lowerBound[1] = minInd - 0.05*step;
+
+  Point upperBound(2);
+  upperBound[0] = values.getDimension() + 0.2 ;
+  upperBound[1] = maxInd + 0.05*step;
+
+  graph.setBoundingBox(Interval(lowerBound, upperBound));
+
+  return graph;
+}
+
+
+/* Draw the HSIC indices */
+Graph HSICEstimatorImplementation::drawHSICIndices() const
+{
+  return drawValues(getHSICIndices(), "HSIC indices");
+}
+
+/* Draw the R2-HSIC indices */
+Graph HSICEstimatorImplementation::drawR2HSICIndices() const
+{
+  return drawValues(getR2HSICIndices(), "R2-HSIC indices");
+}
+
+/* Draw the p-values */
+Graph HSICEstimatorImplementation::drawPValuesPermutation() const
+{
+  return drawValues(getPValuesPermutation(), "p-values by permutation");
+}
+
+/* Get the covariance list */
+PersistentCollection <CovarianceModel> HSICEstimatorImplementation::getCovarianceList() const
+{
+  return covarianceList_;
+}
+
+/* Set the covariance list: dimension is input dimension + 1 */
+void HSICEstimatorImplementation::setCovarianceList(const CovarianceModelCollection & coll)
+{
+  covarianceList_ = coll ;
+  resetIndices();
+}
+
+/* Get the input sample */
+Sample HSICEstimatorImplementation::getInputSample() const
+{
+  return inputSample_;
+}
+
+/* Set the input sample */
+void HSICEstimatorImplementation::setInputSample(const Sample & inputSample)
+{
+  inputSample_ = inputSample;
+  resetIndices();
+}
+
+/* Get the output sample */
+Sample HSICEstimatorImplementation::getOutputSample() const
+{
+  return outputSample_;
+}
+
+/* Set the output sample: must be of dimension one */
+void HSICEstimatorImplementation::setOutputSample(const Sample & outputSample)
+{
+  if(outputSample.getDimension() != 1)
+  {
+    throw NotYetImplementedException(HERE) << "Dimension of output sample should be 1.";
+  }
+
+  outputSample_ = outputSample;
+  resetIndices();
+}
+
+
+/* Reset all indices to void */
+void HSICEstimatorImplementation::resetIndices()
+{
+  HSIC_XY_ = Point();
+  HSIC_XX_ = Point();
+  HSIC_YY_ = Point();
+  R2HSICIndices_ = Point();
+  PValuesPermutation_ = Point();
+}
```

Comment:
```suggestion
  isAlreadyComputedIndices_ = false;
  isAlreadyComputedPValuesPermutation_ = false;
}
```
I also think we should make `PValuesAsymptotic_` and `isAlreadyComputedPValuesAsymptotic_` attributes of `HSICEstimatorImplementation` rather than the derived classes, as this would allow them to be reset together with the other attributes through the `resetIndices` method